### PR TITLE
damlc: check for inconsistent dependencies

### LINKED ISSUE #8553: damlc should error out if `dependencies` have a different LF version
We already have a check that `dependencies` and `data-dependencies` have a newer LF version. For `dependencies` we should error out if they have a different LF version regardless of whether it is newer or older.

### PR DESCRIPTION
**Pull request by robin-da**
This fixes #8553.

We check that all dependencies have the expected LF version.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs [hunk 1]
```diff
@@ -522,18 +525,42 @@ data DependencyInfo = DependencyInfo
...
+        , showDeps inconsistentLfDeps
+        ]
+  where
+    inconsistentLfDeps =
+        filter (\(_, ver) -> ver /= lfTarget) $
+            [ ( (LF.dalfPackageId decodedDalfPkg, decodedUnitId)
+              , (LF.packageLfVersion . LF.extPackagePkg . LF.dalfPackagePkg) decodedDalfPkg
+              )
+            | DecodedDalf{..} <- dalfsFromDependencies
+            , decodedUnitId `elem` mainUnitIds
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs:553]: Let’s turn this into a `Set`. The linear lookup seems like we’re just asking to end up on accidentallyquadratic :)

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs [hunk 2]
```diff
@@ -522,18 +525,42 @@ data DependencyInfo = DependencyInfo
...
+  | null inconsistentLfDeps = Right ()
+  | otherwise = Left $ concat
+        [ "Targeted LF version "
+        , DA.Pretty.renderPretty lfTarget
+        , " but dependencies have different LF versions: "
+        , showDeps inconsistentLfDeps
+        ]
+  where
+    inconsistentLfDeps =
+        filter (\(_, ver) -> ver /= lfTarget) $
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs:548]: can we put the filter in the guard of the list comprehension?

**Review by cocreature — APPROVED** [review]
Nice, thank you!

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs (hunk 1 above)
Comment (robin-da) [reply compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs:553]: You want to turn the `mainUnitIds` field into a `Set` or just here for the lookup?

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs (hunk 2 above)
Comment (robin-da) [reply compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs]: done.

**Pull request merged**